Documentation on how to build navigation

tomjohnson1492 opened the pull request:
I added a documentation page on how to build navigation for your site. This topic is primarily intended for users who have a lot of pages on their site (such as for documentation websites), and want to build a more robust sidebar navigation.

Jekyll combines Liquid with YAML in interesting ways that aren't really documented clearly in the existing docs, except for a brief reference [here](http://jekyllrb.com/docs/datafiles/#the-data-folder). You can read about Liquid on Shopify and YAML in YAML's docs, but exactly how you store YAML files in a Jekyll project and iterate through them using Liquid loops and filters to generate lists of pages is something that isn't clear to a lot of people. (You can see origins of these questions in [previous help issues](https://github.com/jekyll/jekyll-help/issues/266).)  The documentation on navigation would fit well into the Jekyll docs.

@jekyll/documentation

ashmaroli commented:
@DirtyF How about hosting **all `"how to.."`** stuff on another document entirely? (or a subdomain e.g. `tips.jekyllrb.com`?)

/cc @parkr

tomjohnson1492 commented:
Are you proposing that *all* documentation be moved into a subdomain? Or just some types of documentation? 

If it's just *some* types of documentation, my suggestion is that rather than fragmenting the documentation into different domains, we instead create an Advanced page that simply lists the advanced tutorials. 

I get that we probably don't want Jekyll's documentation to seem too complicated. (All product managers want their product to seem simple.) But the result is often a paradox. The more "simple" we try to make something by removing documentation, the harder it can become for people to use the product. As soon as people get passed the first step of installing Jekyll, they're confused at how to go to the next step.

The solution to providing simple instructions for novice users and advanced instructions for advanced users is to separate the info out a bit. Put the advanced content into an advanced area. This is why I proposed adding a link to the navigation document from http://jekyllrb.com/docs/datafiles/. The datafiles doc starts out easy, and when you want to do something more, you transition to more advanced info (progressive disclosure).

But maybe I'm misinterpreting the proposal here. Maybe the subdomain idea is to facilitate something else altogether, such as simplifying merge updates? Would decoupling docs from the other Jekyll code make it easier to deploy updates?

Re tips in the subdomain, there's already http://jekyll.tips/, so that wouldn't work. (Actually, I'd love to see Mike's info be merged with the existing Jekyll doc. I don't know why that's separate in the first place. His tutorials and info are golden.)

The most logical alternative for a subdomain is docs.jekyllrb.com.

ashmaroli commented:
What I'm proposing is that **all** documentation related to non-basic customization of Jekyll be moved to a separate document or subdomain ( new repo ) by itself. `http://tips.jekyllrb.com` was only a suggestion. In theory it will work because `tips` will be a subdomain of `jekyllrb.com`, no way related to `http://jekyll.tips`. But should the idea be considered, we could think of using a different subdomain entirely to avoid confusion with the latter site. 

IMO, `jekyllrb.com` being official documentation should not disseminate tutorials.

pathawks commented:
> What I'm proposing is that all documentation related to non-basic customization of Jekyll be moved to a separate document

I disagree that making it more difficult to find or contribute to documentation would solve any problem.

tomjohnson1492 commented:
I think it's a good idea to have different areas of documentation. I can see value in having a section designated for tutorials, another section designated for reference, another for getting started. And I can see how this topic on implementing navigation would fit into a tutorials section.

But as far as shifting around the different types of content onto different subdomains, I'm not a fan of that. It will make linking more difficult, as well as potentially create confusion among users. 

How about just adding a link to "Tutorials" in the sidebar, and then on that page, add a link to "Implementing Navigation." Or create a _tutorials collection and put them there. Then they could be under jekyllrb.com/tutorials instead of jekyllrb.com/docs. I like that idea.

I have a few other tutorials brewing in the back of my mind. For example, implementing media RSS feeds, converting an HTML theme into a jekyll site, implementing translation, validating links, and more. All of these topics could comfortably fit into a Tutorials collection.

Really, users don't care as long as they can find the information and it answers their question. Good documentation has a mix of reference, tutorials, how-to topics, and more. Think about it -- the foundation of all developer documentation is a "Hello World" tutorial.

ashmaroli commented:
:+1:  to more tutorials.. I would just like them all collected under a single dedicated document/sub-domain.
:+1:  to _tutorials collection folder. It addresses my point with minimal interference.

tomjohnson1492 commented:
Do you want me to create the collection for this? I can. I just wasn't sure if you were waiting for me to do that.

ashmaroli commented on docs/_docs/navigation.md:
```diff
@@ -0,0 +1,570 @@
+{% raw %}<h2>{{ site.data.samplelist.docs_list_title }}</h2>
+<ul>
+   {% for item in site.data.samplelist.docs %}
+   <li><a href="{{ item.url }}" alt="{{ item.title }}">{{ item.title }}</a></li>
```
contents of liquid blocks are indented as well..
```
<ul>
  {% for item in site.data.samplelist.docs %}
    <li><a href="{{ item.url }}" alt="{{ item.title }}">{{ item.title }}</a></li>
  {% endfor %}
</ul>
```

tomjohnson1492 replied:
updated

ashmaroli commented on docs/_docs/navigation.md:
```diff
@@ -0,0 +1,570 @@
+<li><a href="configuration.html" alt="Configuration">Configuration</a></li>
+<li><a href="deployment.html" alt="Deployment">Deployment</a></li>
+</ul>
+</div>
```
indentation required..

tomjohnson1492 replied:
updated

ashmaroli commented on docs/_docs/navigation.md:
````diff
@@ -0,0 +1,570 @@
+{% for item in doclist %}
+<li><a href="{{ item.url }}" alt="{{ item.title }}">{{ item.title }}</a></li>
+{% endfor %}{% endraw %}
+```
````
indentation required..

tomjohnson1492 replied:
updated

tomjohnson1492 commented on docs/_docs/navigation.md:
```diff
@@ -0,0 +1,570 @@
+permalink: /docs/navigation
+---
+
+<style>
```
updated

tomjohnson1492 commented on docs/_docs/navigation.md:
```diff
@@ -0,0 +1,570 @@
+
+<style>
+.result {
+border: 1px solid yellow;
```
updated

tomjohnson1492 commented:
I made the necessary updates around indentation. Submitting for re-review.

ashmaroli commented on docs/_docs/navigation.md:
```diff
@@ -319,19 +323,19 @@ sidebar: toc
-<li><a href="{{ item.url }}">{{ item.title }}</a></li>
-{% endfor %}
+    {% for item in site.data.samplelist[page.sidebar] %}
+    <li><a href="{{ item.url }}">{{ item.title }}</a></li>
```
requires additional indentation here.. :wink:

tomjohnson1492 replied:
updated

ashmaroli commented on docs/_docs/navigation.md:
```diff
@@ -353,8 +357,8 @@ In addition to inserting items from the YAML data file into your list, you also
-<li class="{% if item.url == page.url %}active{% endif %}">
-<a href="{{ item.url }}">{{ item.title }}</a></li>
+    <li class="{% if item.url == page.url %}active{% endif %}">
+    <a href="{{ item.url }}">{{ item.title }}</a></li>
```
- the anchor tag is nested.. so would need to be indented..
- the closing list-item tag on a new-line..

tomjohnson1492 replied:
updated

ashmaroli commented on docs/_docs/navigation.md:
`<style>` => External CSS

DirtyF commented on docs/_sass/_style.scss:
```diff
@@ -1031,3 +1031,11 @@ code.output {
 }
+
+.result {
+  border: 1px solid yellow;
```
I'd rather simply apply `highlight` for now. We can discuss style in a separate PR.

![result-styles](https://cloud.githubusercontent.com/assets/103008/21732810/9fd76e98-d45b-11e6-9e2c-8bb8f3b3a4ce.png)

tomjohnson1492 replied:
Thanks for the comments. I'll make the updates in a couple of days (I'm finishing up some presentations).

tomjohnson1492 replied:
updated, but i did leave the padding style b/c without it, the text is right against the line.

DirtyF commented on docs/_docs/navigation.md:
All these clickable links will generate 404 errors 😭 

Shouldn't we use `#` in href for all examples here?

ashmaroli commented:
Some more ideas:
  - Collect all tutorials under a new **category:** `tutorials`
  - Like how our release news-items are published, allow author to have their name displayed. (optional: link to their blog/website)

/cc @DirtyF, @pathawks

tomjohnson1492 commented on docs/_docs/navigation.md:
updated

tomjohnson1492 commented on docs/_docs/navigation.md:
updated

tomjohnson1492 commented:
I made the updates as noted in the reviews. Submitting for review to merge. Thanks, and sorry for my delay in getting to this.

tomjohnson1492 commented:
@ashmaroli We talked about making tutorials a collection and putting them into another section. I'm guessing I should make that change, right? This will involve me editing the config file and adding more modifications in the nav somewhere.

ashmaroli commented:
I would like that. But I'll leave the making-the-executive-decisions to @DirtyF  :smiley:

parkr commented:
> making tutorials a collection and putting them into another section. I'm guessing I should make that change

@tomjohnson1492 If you're up for it, yes! You can use a lot of the existing templates for `docs`, but change `site.docs` to `site.tutorials` in all the loops and `where` clauses and add `tutorials` to to the config. It may force us to re-think our top-level navigation ("HOME DOCS HELP GITHUB" etc) but for now, I think let's go for it!

tomjohnson1492 commented:
I made an update here that you'll want to look at. Following @parkr's  green-light, I added a new collection called **tutorials** and made it function similar to **docs**. I also added an overview page to the Tutorials section (in addition to the navigation topic that already exists). I think this section will help create a space for a wider number of community contributions -- I can already think of about 5 tutorials I want to add here.

@jekyll/documentation

DirtyF commented on docs/_docs/datafiles.md:
````diff
@@ -71,6 +71,9 @@ You can now render the list of members in a template:
 ```
 
+{: .note .info }
+If your Jekyll site has a lot of pages, such as with documentation websites, see the detailed examples in [how to build robust navigation for your site](../navigation). 
````
We need to update the link to tutorials here.

tomjohnson1492 replied:
ahh, good catch. updated.

DirtyF commented:
@tomjohnson1492  Shouldn't we point to the data documentation reference somewhere in the navigation tutorial?

tomjohnson1492 commented:
I added a reference to the data files topic. Good catch there.

tomjohnson1492 commented:
Just wondering, do you happen to know when this PR will be merged? Does the merging only happen during release windows, or are merges done continuously? My other PR tutorials are dependent on this one.

tomjohnson1492 commented on docs/_includes/primary-nav-items.html:
```diff
@@ -5,6 +5,9 @@
   </li>
+  <li class="{% if page.url contains '/tutorials/' %}current{% endif %}">
+    <a href="/tutorials/home/">Tutorials</a>
+  </li>
```
I could remove the "Tutorials" link from the homepage nav and list it on the ["Help" page](http://jekyllrb.com/help/) right under Documentation. Would that work?

parkr replied:
Yeah, that would work for now 😄

tomjohnson1492 commented:
@parkr I made updates from your review. I moved the "Tutorials" link to the help page and removed the yellow border around the result content.

I'd really like to get this PR merged so that I can add other tutorials here as well. Let me know if I need to do anything else to speed this along. Thanks.

parkr commented:
Let's do it! Thanks 😄 

@jekyllbot: merge +site

parkr commented:
> We are investigating issues serving web hooks as well as assets on GitHub.com.
> – [GitHub Status](https://status.github.com/messages)

@jekyllbot uses webhooks, WOMP

tomjohnson1492 commented:
wow, that was fast. Thanks!
